src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_ioctl.c

```diff
@@ -442,22 +479,76 @@ static bool xocl_xclbin_in_use(struct xocl_dev *xdev)
 	return false;
 }
 
+static int
+xocl_resolver(struct xocl_dev *xdev, struct axlf *axlf, xuid_t *xclbin_id,
+		uint32_t qos,	uint32_t *slot_id)
+{
+	uint32_t s_id = DEFAULT_PL_SLOT;
+	int ret = 0;
+
+	if (!xocl_axlf_section_header(xdev, axlf, SOFT_KERNEL)) {
+		s_id = DEFAULT_PL_SLOT;
+		if (xclbin_downloaded(xdev, xclbin_id, s_id)) {
+			if (qos & XOCL_AXLF_FORCE_PROGRAM) {
+				// We come here if user sets force_xclbin_program
+				// option "true" in xrt.ini under [Runtime] section
+				DRM_WARN("%s Force xclbin download", __func__);
+				*slot_id = s_id;
+			} else {
+				*slot_id = s_id;
+				ret = -EEXIST;
+				goto done;
+			}
+		}
+	}
+	else {
+		static int ps_slot_id = 0;
```

Comment:
static? not sure how does it work in a multi-device system, all driver instances share this variable? shouldn't xdev holds all the per-device context?

Reply:
Yes. For multiple device this will not work. Moved this to now as a device specific filed